src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -0,0 +1,447 @@
+        builder.add(str_token, offset);
+        offset += encoding_lengths[current_index++];
+    }
+
+    builder.build();
+    write_buf.finalize();
+
+    /// Write FST size
+    writeVarUInt(buffer.size(), *term_dict_file_stream);
+    current_segment.item_dict_start_offset += getLengthOfVarUInt(buffer.size());
+
+    /// Write FST content
+    term_dict_file_stream->write(reinterpret_cast<char*>(buffer.data()), buffer.size());
+    current_segment.item_dict_start_offset += buffer.size();
+
+    current_size = 0;
+    current_postings.clear();
+    current_segment.segment_id = getNextSegmentID();
+
+    segment_file_stream->sync();
+    term_dict_file_stream->sync();
+    postings_file_stream->sync();
+}
+
+void GinIndexStoreReader::init_file_streams()
+{
+    String segment_file_name = store->getName() + ".gin_seg";
+    String item_dict_file_name = store->getName() + ".gin_dict";
+    String postings_file_name = store->getName() + ".gin_post";
+
+    segment_file_stream = store->storage->readFile(segment_file_name, {}, std::nullopt, std::nullopt);
+    term_dict_file_stream = store->storage->readFile(item_dict_file_name, {}, std::nullopt, std::nullopt);
+    postings_file_stream = store->storage->readFile(postings_file_name, {}, std::nullopt, std::nullopt);
+}
+void GinIndexStoreReader::readSegments()
+{
+    GinIndexSegments segments;
+
+    auto segment_num = store->getSegmentNum();
+    if (segment_num == 0)
+        return;
+
+    segments.assign(segment_num, {});
+
+    if (segment_file_stream == nullptr)
+    {
+        init_file_streams();
+    }
+    segment_file_stream->read(reinterpret_cast<char*>(segments.data()), segment_num * sizeof(GinIndexSegment));
+    for (size_t i = 0; i < segment_num; ++i)
+    {
+        auto seg_id = segments[i].segment_id;
+        auto term_dict = std::make_shared<TermDictionary>();
+        term_dict->postings_start_offset = segments[i].postings_start_offset;
+        term_dict->item_dict_start_offset = segments[i].item_dict_start_offset;
+        store->term_dicts[seg_id] = term_dict;
+    }
+}
+
+void GinIndexStoreReader::readTermDictionary(UInt32 segment_id)
+{
+    /// Check validity of segment_id
+    auto it{ store->term_dicts.find(segment_id) };
+    if (it == store->term_dicts.cend())
+    {
+        throw Exception("Invalid segment id " + std::to_string(segment_id), ErrorCodes::LOGICAL_ERROR);
+    }
+
+    it->second->offsets.getData().clear();
+
+    /// Set file pointer of term dictionary file
+    term_dict_file_stream->seek(it->second->item_dict_start_offset, SEEK_SET);
+
+    /// Read FST size
+    size_t fst_size{0};
+    readVarUInt(fst_size, *term_dict_file_stream);
+
+    /// Read FST content
+    it->second->offsets.getData().resize(fst_size);
+    term_dict_file_stream->readStrict(reinterpret_cast<char*>(it->second->offsets.getData().data()), fst_size);
+}
+
+SegmentedPostingsListContainer GinIndexStoreReader::readSegmentedPostingsLists(const String& token)
```

Comment:
As class `GinIndexStoreReader` is mostly concerned with reading the GIN index from persistence into memory, it might as well be called `GinIndexStoreDeserializer`. `Reader` gives the false impression that this class also handles reading the in-memory index data.

Reply:
Fixed.